### PR TITLE
WIP: corrupted isbits Union fields when uninitialized

### PR DESCRIPTION
**Pull request by quinnj**
The minimum repro is
```julia
v = Vector{Float64}(8000000);
mutable struct UnionField2
    x::Union{Void, Int}
    UnionField2() = new()
end
UnionField2().x
```
I think this is what's needed in `datatype.c`? But I'm not sure what/where the corresponding fix would be in `codegen.cpp`; @vtjnash could you take a look? Also happy to take a stab if you could point me to the right place.

### CODE REVIEW

File: src/datatype.c [hunk 1]
```diff
@@ -735,6 +735,11 @@ JL_DLLEXPORT jl_value_t *jl_new_structv(jl_datatype_t *type, jl_value_t **args,
         if (jl_field_isptr(type, i)) {
             *(jl_value_t**)((char*)jl_data_ptr(jv)+jl_field_offset(type,i)) = NULL;
         }
+        jl_value_t *ft = jl_field_type(type, i);
```
Comment (vtjnash) [inline src/datatype.c]: should only check this if `!isptr`

**Comment by vtjnash** [comment]
The equivalent codegen is https://github.com/JuliaLang/julia/blob/9c5ee20a8112d6999918ef992fb93460aa028de5/src/cgutils.cpp#L2439-L2457

and needs to run from `nargs` to `nf`
```cpp
        for (size_t i = nargs; i < nf; i++) {
            if (!jl_field_isptr(sty, i) && jl_is_uniontype(jl_field_type(sty, i))
                tbaa_decorate(tbaa_unionselector, ctx.builder.CreateStore(
                        ConstantInt::get(T_int8, 0),
                        ctx.builder.CreateGEP(emit_bitcast(ctx, strct, T_pint8),
                            ConstantInt::get(T_size, jl_field_offset(sty, i) + jl_field_size(sty, i) - 1)));

```
(the isbits case case several lines up will need a similar adjustment to extend its for-loop from na to nf and handle inserting the missing uninitialized union bits)

File: test/core.jl [hunk 2]
```diff
@@ -5505,6 +5505,33 @@ x.u = initvalue2(Base.uniontypes(U)[1])
...
+    UnionField4() = new()
+end
+@test UnionField4().x === nothing
+
+struct UnionField5
+    x::Union{Void, Float64}
+    y::Union{Void, Int8}
+    UnionField5() = new()
+end
+@test UnionField5().x === nothing
```
Comment (quinnj) [inline test/core.jl]: @vtjnash, this is the only case that still fails at the REPL (I can't trigger a failure when run in a script). Are there any other places that we need to update? Something separate for struct/mutable struct in datetype.c?

Comment (vtjnash) [reply test/core.jl]: Looks like an LLVM bug (https://bugs.llvm.org/show_bug.cgi?id=34940)

Comment (quinnj) [reply test/core.jl]: Ah......so what should we do here? Merge and open an issue to track upstream? Is there some kind of patch we could carry locally? (I say that realizing it's way over my head though in terms of what a fix here would entail)

Comment (quinnj) [reply test/core.jl]: Bump; thoughts on what to do here?

Comment (vtjnash) [reply test/core.jl]: If these tests are working (which they seem to be), we can go ahead and merge this. The bug seems to be pretty serious, and is largely unrelated to the union optimization work. But we need upstream to fix that.

Comment (quinnj) [reply test/core.jl]: Sounds good. I'll open an issue so we can track the upstream fix (so we don't forget)

**Pull request merged**